packages/elasticsearch-asset-apis/src/elasticsearch-reader-api/ElasticsearchReaderAPI.ts

```diff
@@ -111,21 +112,100 @@ export class ElasticsearchReaderAPI {
+     *   https://github.com/terascope/elasticsearch-assets/issues/948
+     */
     async fetch(queryParams: ReaderSlice = {}): Promise<DataEntity[]|DataFrame|Buffer> {
-        // attempt to get window if not set
+        const countExpansionFactor = 1.5;
+        let querySize = 10000;
+        const retryLimit = 5;
+
+        // attempt to get window if not set (sets this.windowSize as side effect)
         if (!this.windowSize) await this.setWindowSize();
 
-        // if we did go ahead and complete query
-        const query = buildQuery(this.config, {
-            ...queryParams, count: this.windowSize
-        });
+        // set querySize to the lesser of expandedSize or windowSize
+        if (queryParams.count) {
+            const expandedSize = Math.ceil(queryParams.count * countExpansionFactor);
+            if (this.windowSize) {
+                if (expandedSize >= this.windowSize) {
+                    throw new Error(`The query size, ${expandedSize}, is greater than the index.max_result_window: ${this.windowSize}`);
+                } else {
+                    querySize = expandedSize;
+                }
+            }
+        }
 
-        return this.client.search(
-            query,
-            this.config.response_type ?? FetchResponseType.data_entities,
-            this.config.type_config
+        const _fetch = async ():
+        Promise<DataEntity[]|DataFrame|Buffer> => {
+            const query = buildQuery(this.config, {
+                ...queryParams, count: querySize
+            });
+
+            const result = await this.client.search(
+                query,
+                this.config.response_type ?? FetchResponseType.data_entities,
+                this.config.type_config
+            );
+
+            const resultSize = this._getResultSize(result);
+
+            if (resultSize === querySize) {
+                // it's unlikely that this condition would be true without
+                // resultSize actually being larger than querySize so we will
+                // throw away these results, expand querySize and query again
+                // by relying on pRetry
+                const expandedSize = Math.ceil(querySize * countExpansionFactor);
+                if (this.windowSize) {
+                    if (expandedSize >= this.windowSize) {
+                        throw new Error(`The query size, ${expandedSize}, is greater than the index.max_result_window: ${this.windowSize}`);
+                    } else {
+                        querySize = expandedSize;
+                    }
+                }
+                const msg = `The result set contained exactly ${resultSize} records, searching again with size: ${expandedSize}`;
+                this.logger.debug(msg);
+                throw new Error(msg); // throw for pRetry
+            }
+
+            return result;
+        };
+
+        const result = await pRetry(() => _fetch(),
+            {
+                backoff: 1.1,
```

Comment:
The settings on the `pRetry` are meant to speed things up.  In this case the failure is not due to a service problem or temporary network condition, but instead comes from an assumption that has become untrue, so we retry fairly quickly, but let it dial back a little to avoid overloading the cluster.